usbserial/src/main/java/com/felhr/usbserial/SerialInputStream.java

```diff
@@ -54,6 +54,18 @@ public int read(byte[] b)
         return device.syncRead(b, timeout);
     }
 
+    @Override
+    public int read(byte b[], int off, int len)
+    {
+        if (off == 0 && len == b.length) {
+            return read(b);
+        }
+        byte[] slice = new byte[len];
```

Comment:
Not ideal - the repeated slice creation will create lots of collectable garbage.
Suggestion:
- create a `device.syncRead()` function which contains an offset into the array
- inside `syncRead()`, use the array-offset version of `UsbDeviceConnection.bulkTransfer()`

This avoids repeated slice allocations anywhere in the syncRead() call chain.

Reply:
Probably the best way but I see some problems:
- array-offset version of bulkTransfer is added in api 18. Not a terrible issue because most people should not be using Android 4.x in the first place but it would force us to add two implementations if we want to keep the oldest versions supported.
- FTDI devices use their own implementation of syncRead and have to be taken into account.